astropy/coordinates/baseframe.py

```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
         return out, converted
 
 
+class CartesianRepresentationFrameAttribute(FrameAttribute):
+    """
+    A frame attribute that is a CartesianRepresentation with specified units.
+
+    Parameters
+    ----------
+    default : object
+        Default value for the attribute if not provided
+    secondary_attribute : str
+        Name of a secondary instance attribute which supplies the value if
+        ``default is None`` and no value was supplied during initialization.
+    unit : unit object or None
+        Name of a unit that the input will be converted into. If None, no
+        unit-checking or conversion is performed
+    """
+    def __init__(self, default=None, secondary_attribute='', unit=None, shape=None):
+        super(CartesianRepresentationFrameAttribute, self).__init__(default, secondary_attribute)
+        self.unit = unit
+
+    def convert_input(self, value):
+        """
+        Checks that the input is a CartesianRepresentation with the correct unit, or
+        the special value ``0``.
+
+        Parameters
+        ----------
+        value : object
+            Input value to be converted.
+
+        Returns
+        -------
+        out, converted : correctly-typed object, boolean
+            Tuple consisting of the correctly-typed object and a boolean which
+            indicates if conversion was actually performed.
+
+        Raises
+        ------
+        ValueError
+            If the input is not valid for this attribute.
+        """
+        if np.all(value == 0) and self.unit is not None:
+            return CartesianRepresentation(np.zeros(3) * self.unit), True
+        else:
+            converted = True
+            # if it's a CartesianRepresentation, get the xyz Quantity
+            value = getattr(value, 'xyz', value)
+            if not (hasattr(value, 'unit') ):
+                raise TypeError('Tried to set a CartesianRepresentationFrameAttribute with '
+                                'something that does not have a unit.')
+            oldvalue = value
+            value = u.Quantity(oldvalue, self.unit, copy=False)
+            if (oldvalue.unit == value.unit and hasattr(oldvalue, 'value') and
+                np.all(oldvalue.value == value.value)):
+                converted = False
```

Comment:
As I understand `FrameAttribute`, converted should be true if any work has been done, so that the "reworked" value can be stored and retrieved later without again converting. Since here you always create a new `CartesianRepresentation`, this suggests `convert` should always be `True`. (@taldcroft: since you wrote this code: is this correct?)

If the above holds, it might be an idea to check whether a given value already is a `CartesianRepresentation` with the correct unit, since that would be the only case for which no conversion would be necessary.